Powerwall: Reuse authentication cookie

choumarin opened the pull request:
Store and reuse the authentication cookie to the Powerwall.

For multiple users (including me) the new password is only valid for a limited time, but the cookie duration seems unlimited. This will help with #129440 as HA reboots will not require resetting the password.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #129440
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

choumarin committed:
Reuse authentication cookie

Store and reuse the authentication cookie to the powerwall.
This will help with #129440

bdraco reviewed (changes requested):
Thanks for the PR. Please add test coverage for the new code.

choumarin committed:
Handle reauth and add test

bdraco commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -116,14 +130,36 @@ async def _update_data(self) -> PowerwallData:
+    if auth_cookie_value:
+        cookie_jar.update_cookies(
+            {AUTH_COOKIE_KEY: auth_cookie_value},
+            URL("http://" + entry.data[CONF_IP_ADDRESS]),
```
```suggestion
            URL(f"http://{entry.data[CONF_IP_ADDRESS]}"),
```

choumarin replied:
Done

bdraco commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -147,7 +182,6 @@ async def async_setup_entry(hass: HomeAssistant, entry: PowerwallConfigEntry) ->
             )
             return False
         except AccessDeniedError as err:
-            _LOGGER.debug("Authentication failed", exc_info=err)
```
Why drop this debug logging?

choumarin replied:
sorry, removed too many temporary logs I had added
reverted

bdraco commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -163,7 +197,16 @@ async def async_setup_entry(hass: HomeAssistant, entry: PowerwallConfigEntry) ->
+        cookie_jar,
+        entry,
+        ip_address,
+        entry.data.get(CONF_PASSWORD),
```
Why change this from `password`?

choumarin replied:
because password might have been updated to None to use the cookie, however, we still want to keep the real password for _recreate_powerwall_login

bdraco replied:
If we clear it above, it means `_recreate_powerwall_login` will always try with `""` and fail.

I think we should still pass it in, as well as a flag to try cookie auth first, and than fallback to password if it doesn't work

choumarin replied:
Pushed a new commit with several tests. Also fix proper handling of reauth flow by clearing the cookie.
To sum up:
- init will try to use the cookie if one is present
- if there is an access denied, _recreate_powerwall_login will use the password to get a new cookie
- if _recreate_powerwall_login fails as well, it goes to the reauth flow
- the reauth flow will ask password from user and use that only. It will also clear the cookie so that a new one is obtained in init

choumarin committed:
PR comments

choumarin committed:
Clean up and add tests

Clearer distinction when using or not the auth cookie.
Several tests around proper usage of auth cookie.

bdraco commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -116,14 +130,37 @@ async def _update_data(self) -> PowerwallData:
                 return data
         raise RuntimeError("unreachable")
 
+    async def save_auth_cookie(self) -> None:
```
```suggestion
    @callback
    def async_save_auth_cookie(self) -> None:
```

Nothing is being awaited here, this can be marked as an async safe callback function instead and the `await` removed

bdraco commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -163,7 +200,16 @@ async def async_setup_entry(hass: HomeAssistant, entry: PowerwallConfigEntry) ->
+        password,
+        runtime_data,
+    )
+    await manager.save_auth_cookie()
```
This seems unneeded since no request has been made to login yet and we would only re-save the existing cookie we feed in above

choumarin replied:
Maybe I'm misreading, but we connect above on L174 (on latest commit) with _login_and_fetch_base_info

bdraco replied:
Will double check it tomorrow, after some sleep, it's getting late here

bdraco replied:
I agree. It possible for it to be updated there.  We should guard the config entry update so it doesn't update anything if nothing has changed.

bdraco replied:
nevermind async_update_entry already checks that. so all good here.

bdraco commented on homeassistant/components/powerwall/config_flow.py:
```diff
@@ -257,6 +257,8 @@ async def async_step_reauth_confirm(
+                # We have a new valid connection, old cookie is no longer valid
+                user_input[CONFIG_ENTRY_COOKIE] = None
                 return self.async_update_reload_and_abort(
                     reauth_entry, data_updates=user_input
```
```suggestion
                    reauth_entry, data_updates={**user_input, CONFIG_ENTRY_COOKIE: None}
```
This would be a bit cleaner than mutate user_input

bdraco commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -116,14 +130,37 @@ async def _update_data(self) -> PowerwallData:
+                    self.entry,
+                    data={**self.entry.data, CONFIG_ENTRY_COOKIE: cookie.value},
+                )
+                _LOGGER.debug("Saved auth cookie")
```
```suggestion
                _LOGGER.debug("Saved auth cookie")
                break
```

bdraco replied:
No need to keep iterating once we find it

bdraco commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -132,7 +169,7 @@ async def async_setup_entry(hass: HomeAssistant, entry: PowerwallConfigEntry) ->
         try:
             base_info = await _login_and_fetch_base_info(
-                power_wall, ip_address, password
+                power_wall, ip_address, password, use_auth_cookie
```
If someone factory resets the gateway and the cookie is saved but no longer valid, it looks like it will never try to login with this change

choumarin replied:
Last commit should handle that

choumarin committed:
Apply suggestions from code review

Co-authored-by: J. Nick Koston <nick@koston.org>

choumarin committed:
Retry init with password if cookie fails

Includes test and PR comments

bdraco committed:
Merge branch 'dev' into choumarin/powerwall_reuse_cookie

bdraco commented:
Looks good. Testing on production now

bdraco reviewed (approved):
Thanks @choumarin 

Tested on production. All seems well

choumarin commented:
Thanks you!
